test/riak_kv_sweeper_SUITE.erl

```diff
@@ -0,0 +1,892 @@
+    ThrottleExtraMsecs = expected_throttle_extra_msecs(
+                           NumMutatedKeys, ThrottleWaitMsecs),
+
+    {ok, ThrottleDiv} = application:get_env(riak_kv, sweep_window_throttle_div),
+    case ThrottleDiv > ThrottleWaitMsecs of
+        true ->
+            0;
+        false ->
+            (ThrottleMsecs + ThrottleExtraMsecs) div ThrottleDiv
+    end.
+
+expected_pace_throttle_total_msecs(NumKeys, NumMutatedKeys, NumKeysPace, ThrottleWaitMsecs) ->
+    ThrottleMsecs =
+        (NumKeys div NumKeysPace) * ThrottleWaitMsecs +
+        (NumMutatedKeys div NumKeysPace) * ThrottleWaitMsecs,
+    {ok, ThrottleDiv} = application:get_env(riak_kv, sweep_window_throttle_div),
+    case ThrottleDiv > ThrottleWaitMsecs of
+        true ->
+            0;
+        false ->
+            ThrottleMsecs div ThrottleDiv
+    end.
+
+expected_throttle_extra_msecs(NumMutatedKeys, WaitTimeMsecs) ->
+    (NumMutatedKeys div 100) * WaitTimeMsecs.
+
+expected_throttle_msecs(NumKeys, ThrottleAfterBytes, WaitTimeMsecs, ObjSizeBytes) ->
+    case ThrottleAfterBytes div ObjSizeBytes of
+        0 ->
+            NumKeys * WaitTimeMsecs;
+        1 ->
+            (NumKeys div 2)  * WaitTimeMsecs;
+        _ ->
+            (NumKeys div (ThrottleAfterBytes div ObjSizeBytes))  * WaitTimeMsecs
+    end.
+
+riak_object_bin(B, K, ObjSizeBytes) ->
+    riak_object:to_binary(v1, riak_object:new(B, K, <<0:ObjSizeBytes/unit:8>>)).
+
+add_hours(Hour, Inc)  ->
+    (Hour + Inc) rem 24.
+
+%% Waiting for 2500 msecs because at least 1 second must elapse
+%% before a sweep is re-run using the run_interval (now - ts) > 1,
+%% see kv_sweeper:expired
+min_scheduler_response_time_msecs() ->
+    2000.
+
+pick(List) when length(List) > 0 ->
+    N = random:uniform(length(List)),
+    lists:nth(N, List).
+
+sweep_all_indices(Indices) ->
+    sweep_all_indices(Indices, ok).
+
+sweep_all_indices(Indices, ExpectedResult) ->
+    sweep_all_indices(Indices, ExpectedResult, sweeper_callback_1).
+
+sweep_all_indices(Indices, ExpectedResult, SweepParticipant) ->
+    riak_kv_sweeper:enable_sweep_scheduling(),
+    tick_per_index(Indices),
+    ReceiveFun = fun(I) ->
+                         ExpectedResult = receive_msg({ok, successful_sweep, SweepParticipant, I})
+                 end,
+    lists:foreach(ReceiveFun, Indices).
+
+tick_per_index(Indices) ->
+    application:set_env(riak_kv, sweep_concurrency, length(Indices)),
+    %% Only one sweep scheduled per tick, so send a tick for each index:
+    lists:foreach(fun(_) -> riak_kv_sweeper:sweep_tick() end, Indices).
+
+assert_all_indices_idle(Indices) ->
+    {_, Sweeps} = riak_kv_sweeper:status(),
+    SweepData = [{riak_kv_sweeper_state:sweep_state(S),
+                  riak_kv_sweeper_state:sweep_index(S)} ||
+                    S <- Sweeps],
+    SweepIndices = lists:sort([I || {idle, I} <- SweepData]),
+    SweepIndices = Indices.
+
+receive_msg(Msg) ->
+    ok = match_retry(
```

Comment:
Boy, that's pretty weird, and definitely makes me uneasy. I guess if it works around the issue then I'm okay with putting this code in, but I would definitely add some comments around it explaining why we're doing this, otherwise this code looks very strange and confusing to me.